Aegis v0.4.3 - The application suddenly stops when adding images

# Actual behaviour
Supposedly, when I scan an image, then add an image, this application runs well, and that ability can be done.

# Expected behaviour
This application suddenly stops when I add a picture.  Even though I have tried doing it repeatedly, the same problem still occurs.

# Step to reproduce
- Install and open the application. 

- In the security section, select "Password".

- Enter the password, confirm the password.  Then click to continue.

- Click the plus icon, then select the "Scan image" feature.

- Select the image you want to add.  

- Then you will find problems like I have experienced.

# Information
- Device: Nubia n2 lite
- Android version: 8.1.0
- Application version: 0.4.3

# Recording of the bug
https://youtu.be/SQ6TvuVlMgY

Do you mean logcat below?

<code>FATAL EXCEPTION: main
07-03 19:35:49.664 22470 22470 E AndroidRuntime: Process: com.beemdevelopment.aegis, PID: 22470
07-03 19:35:49.664 22470 22470 E AndroidRuntime: java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=7, result=-1, data=Intent { dat=content://media/external/images/media/182027 }} to activity {com.beemdevelopment.aegis/com.beemdevelopment.aegis.ui.MainActivity}: java.lang.SecurityException: Permission Denial: reading com.android.providers.media.MediaProvider uri content://media/external/images/media/182027 from pid=22470, uid=10218 requires android.permission.READ_EXTERNAL_STORAGE, or grantUriPermission()
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.app.ActivityThread.deliverResults(ActivityThread.java:4322)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.app.ActivityThread.handleSendResult(ActivityThread.java:4366)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.app.ActivityThread.-wrap19(Unknown Source:0)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1679)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.os.Handler.dispatchMessage(Handler.java:106)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.os.Looper.loop(Looper.java:164)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.app.ActivityThread.main(ActivityThread.java:6550)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at java.lang.reflect.Method.invoke(Native Method)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:869)
07-03 19:35:49.664 22470 22470 E AndroidRuntime: Caused by: java.lang.SecurityException: Permission Denial: reading com.android.providers.media.MediaProvider uri content://media/external/images/media/182027 from pid=22470, uid=10218 requires android.permission.READ_EXTERNAL_STORAGE, or grantUriPermission()
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.os.Parcel.readException(Parcel.java:2005)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:183)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.database.DatabaseUtils.readExceptionWithFileNotFoundExceptionFromParcel(DatabaseUtils.java:146)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.content.ContentProviderProxy.openTypedAssetFile(ContentProviderNative.java:698)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.content.ContentResolver.openTypedAssetFileDescriptor(ContentResolver.java:1410)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.content.ContentResolver.openAssetFileDescriptor(ContentResolver.java:1247)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.content.ContentResolver.openInputStream(ContentResolver.java:967)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at com.beemdevelopment.aegis.ui.MainActivity.onScanImageResult(MainActivity.java:274)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at com.beemdevelopment.aegis.ui.MainActivity.onActivityResult(MainActivity.java:183)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.app.Activity.dispatchActivityResult(Activity.java:7291)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     at android.app.ActivityThread.deliverResults(ActivityThread.java:4318)
07-03 19:35:49.664 22470 22470 E AndroidRuntime:     ... 9 more</code>

--- a/app/src/main/java/com/beemdevelopment/aegis/ui/MainActivity.java
+++ b/app/src/main/java/com/beemdevelopment/aegis/ui/MainActivity.java
@@ -65,6 +65,7 @@ public class MainActivity extends AegisActivity implements EntryListView.Listene
 
     // permission request codes
     private static final int CODE_PERM_CAMERA = 0;
+    private static final int CODE_PERM_READ_STORAGE = 1;
 
     private AegisApplication _app;
     private DatabaseManager _db;
@@ -106,12 +107,7 @@ public class MainActivity extends AegisActivity implements EntryListView.Listene
         });
         findViewById(R.id.fab_scan_image).setOnClickListener(view -> {
             _fabMenu.collapse();
-
-            Intent galleryIntent = new Intent(Intent.ACTION_PICK);
-            galleryIntent.setDataAndType(android.provider.MediaStore.Images.Media.INTERNAL_CONTENT_URI, "image/*");
-
-            Intent chooserIntent = Intent.createChooser(galleryIntent, getString(R.string.select_picture));
-            startActivityForResult(Intent.createChooser(chooserIntent, getString(R.string.select_picture)), CODE_SCAN_IMAGE);
+            startScanImageActivity();
         });
         findViewById(R.id.fab_scan).setOnClickListener(view -> {
             _fabMenu.collapse();
@@ -194,6 +190,9 @@ public class MainActivity extends AegisActivity implements EntryListView.Listene
         switch (requestCode) {
             case CODE_PERM_CAMERA:
                 startScanActivity();
+                break;
+            case CODE_PERM_READ_STORAGE:
+                startScanImageActivity();
                 break;
         }
     }
@@ -369,6 +368,18 @@ public class MainActivity extends AegisActivity implements EntryListView.Listene
 
         Intent scannerActivity = new Intent(getApplicationContext(), ScannerActivity.class);
         startActivityForResult(scannerActivity, CODE_SCAN);
+    }
+
+    private void startScanImageActivity() {
+        if (!PermissionHelper.request(this, CODE_PERM_READ_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE)) {
+            return;
+        }
+
+        Intent galleryIntent = new Intent(Intent.ACTION_PICK);
+        galleryIntent.setDataAndType(android.provider.MediaStore.Images.Media.INTERNAL_CONTENT_URI, "image/*");
+
+        Intent chooserIntent = Intent.createChooser(galleryIntent, getString(R.string.select_picture));
+        startActivityForResult(Intent.createChooser(chooserIntent, getString(R.string.select_picture)), CODE_SCAN_IMAGE);
     }
 
     private void doShortcutActions() {

